cpp/src/arrow/types/string.h

```diff
@@ -34,85 +34,141 @@ namespace arrow {
+      const ArrayPtr& values, int32_t null_count = 0,
+      const std::shared_ptr<Buffer>& null_bitmap = nullptr)
+      : StringArray(type, length, offsets, values, null_count, null_bitmap),
+        char_type(std::dynamic_pointer_cast<CharType>(type).get()) {}
+
+  // The fixed size of each string
+  // (TODO), right now this is in bytes, DB systems seems to
+  // generally store this as characters.  Which would mean 3-6x the number of bytes
+  // stored)
+  int32_t string_size() const { return char_type->size; }
+  Status Validate() const override;
+
  private:
-  UInt8Array* bytes_;
-  const uint8_t* raw_bytes_;
+  // for convenience
+  CharType* char_type;
 };
 
-// String builder
-class StringBuilder : public ListBuilder {
+// BinaryBuilder : public ListBuilder
+class BinaryBuilder : public ListBuilder {
  public:
-  explicit StringBuilder(MemoryPool* pool, const TypePtr& type)
+  explicit BinaryBuilder(MemoryPool* pool, const TypePtr& type)
       : ListBuilder(pool, std::make_shared<UInt8Builder>(pool, value_type_), type) {
     byte_builder_ = static_cast<UInt8Builder*>(value_builder_.get());
   }
 
+  Status Append(const uint8_t* value, int32_t length) {
+    RETURN_NOT_OK(ListBuilder::Append());
+    return byte_builder_->Append(value, length);
+  }
+
+  std::shared_ptr<Array> Finish() override {
+    return ListBuilder::Transfer<BinaryArray>();
+  }
+
+ protected:
+  UInt8Builder* byte_builder_;
+  static TypePtr value_type_;
+};
+
+// String builder
+class StringBuilder : public BinaryBuilder {
+ public:
+  explicit StringBuilder(MemoryPool* pool, const TypePtr& type)
+      : BinaryBuilder(pool, type) {}
+
   Status Append(const std::string& value) { return Append(value.c_str(), value.size()); }
 
   Status Append(const char* value, int32_t length) {
-    RETURN_NOT_OK(ListBuilder::Append());
-    return byte_builder_->Append(reinterpret_cast<const uint8_t*>(value), length);
+    return BinaryBuilder::Append(reinterpret_cast<const uint8_t*>(value), length);
   }
+
   Status Append(const std::vector<std::string>& values, uint8_t* null_bytes);
 
   std::shared_ptr<Array> Finish() override {
     return ListBuilder::Transfer<StringArray>();
   }
+};
 
- protected:
-  UInt8Builder* byte_builder_;
+// Fixed width char builder
+class CharBuilder : public StringBuilder {
+ public:
+  explicit CharBuilder(MemoryPool* pool, const std::shared_ptr<CharType>& type)
+      : StringBuilder(pool, std::static_pointer_cast<DataType>(type)) {
+    DCHECK(type);
+    fixed_length_ = type->size;
+  }
 
-  static TypePtr value_type_;
+  Status Append(const std::string& value) { return Append(value.c_str(), value.size()); }
+
+  Status Append(const char* value, int32_t length) {
+    if (length != fixed_length_) {
+      return Status::Invalid("Invalid length specified for fixed character type");
+    }
+    return BinaryBuilder::Append(reinterpret_cast<const uint8_t*>(value), length);
```

Comment:
This brings up another question that may be worth discussing on the ML.

Parquet has the notion of BYTE_ARRAY (here: Binary) and FIXED_LEN_BYTE_ARRAY (missing from Arrow at the moment). In the latter case, you store the length in the type metadata, and then it's not necessary to encode the length (i.e. we could omit the offsets here).